### PR TITLE
Move AStatement mutable members into isolated Input/Output objects

### PR DESCRIPTION
**Pull request by jdconrad**
This is the follow up to #53075, isolating the mutable data on the statement nodes as the referenced change did the expression nodes.

This is a step toward the long-term goal of making the "user" trees nodes immutable. This change isolates the mutable data for statement nodes in the "user" tree during the semantic (analysis) phase by moving the mutable data into Input and Output objects. These objects are created locally during the semantic phase to share information required for semantic checking between parent-child nodes.

Note that for this change, Input and Output are still stored in a mutable way on the statement nodes. This will not be the case once the semantic (analysis) phase and ir generation (write) phase are combined in a future change.

Relates #49869

### CODE REVIEW

**Comment by jdconrad** [comment]
@elasticmachine test elasticsearch-ci/2

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/2

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AStatement.java [hunk 1]
```diff
@@ -30,66 +30,76 @@
...
+         */
+        int statementCount = 0;
+    }
 
-    /**
-     * Set to the approximate number of statements in a loop block to prevent
-     * infinite loops during runtime.
-     */
-    int statementCount = 0;
+    // TODO: remove placeholders once analysis and write are combined into build
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AStatement.java]: Can you ref issues in TODO's like this? eg `TODO: ..., see #12345`.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SExpression.java [hunk 2]
```diff
@@ -44,36 +44,40 @@ public SExpression(Location location, AExpression expression) {
     }
 
     @Override
-    void analyze(ScriptRoot scriptRoot, Scope scope) {
+    Output analyze(ScriptRoot scriptRoot, Scope scope, Input input) {
+        this.input = input;
+        output = new Output();
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SExpression.java]: Consider creating output close to it's first use, at line 68.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SFor.java [hunk 3]
```diff
@@ -52,12 +52,15 @@ public SFor(Location location, ANode initializer, AExpression condition, AExpres
     }
 
     @Override
-    void analyze(ScriptRoot scriptRoot, Scope scope) {
+    Output analyze(ScriptRoot scriptRoot, Scope scope, Input input) {
+        this.input = input;
+        output = new Output();
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SFor.java]: Same as above, consider changing where this creation occurs.  It's about ~50 lines away from actual use.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review! Will address your comments soon.

**Review by rjernst — APPROVED** [review]

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AStatement.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AStatement.java]: https://github.com/elastic/elasticsearch/issues/53561

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SExpression.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SExpression.java]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SFor.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SFor.java]: Done.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review. Will commit as soon as CI completes.

**Pull request merged**